### PR TITLE
Run nodejs stack as node user rather than root

### PR DESCRIPTION
**Pull request by Kamran64**
### Checklist:

- [x] Read the [Code of Conduct](https://github.com/appsody/website/blob/master/CODE_OF_CONDUCT.md) and [Contributing Guidelines](https://github.com/appsody/website/blob/master/CONTRIBUTING.md).

- [x] Followed the [commit message guidelines](https://github.com/appsody/website/blob/master/CONTRIBUTING.md#commit-message-guidelines).

- [x] Stack adheres to [Appsody stack structure](https://github.com/appsody/website/blob/master/content/docs/stacks/stacks-overview.md#stack-structure).

### Modifying an existing stack:

- [x] Updated the stack version in `stack.yaml`

- Run the nodejs stack as node user rather than root

### Related Issues:
Related to https://github.com/appsody/stacks/issues/518

### CODE REVIEW

**Comment by Kamran64** [comment]
@sam-github - Might be better to handle this stack first before nodejs-express. Need to test how these changes will impact other downstream stacks too e.g. express, functions, loopback.

**Comment by sam-github** [comment]
This fixes #566

File: incubator/nodejs/image/Dockerfile-stack [hunk 1]
```diff
@@ -1,5 +1,5 @@
 FROM node:12
-
+
```
Comment (sam-github) [inline incubator/nodejs/image/Dockerfile-stack]: nit: this looks like unintentional addition of whitespace

**Comment by sam-github** [comment]
@skoh7645 You asked:

> Can we turn off writing to the resourceRegistry.log file as users would not see this anyway?

No.
￼
￼> If not, could we write the file elsewhere?

No.

I have opened an issue on ibmapm-restclient.

**Comment by sam-github** [comment]
@skoh7645 You mentioned an alternative technique `ENV APPSODY_USER_RUN_AS_LOCAL=true`, I tried:
1. added that to incubator/nodejs-express/image/Dockerfile-stack,
2. ran `appsody stack package`,
3. `mkdir ex; cd ex; appsody init dev.local/nodejs-express`,
4. `appsody run`,

and hit the same

> EACCES: permission denied, open '/project/resourceRegistry.log

I expected to see what you reported:
```
npm ERR! correctMkdir failed to make directory /.npm/_locks
[Container] npm ERR! code EACCES
[Container] npm ERR! syscall mkdir
[Container] npm ERR! path /.npm
[Container] npm ERR! errno -13
```

So I'm doing something wrong. Would you like to give me repro information?

```
nodejs-express/ex (pr/604 *$%) % appsody version
appsody 0.5.3
nodejs-express/ex (pr/604 *$%) % uname -a
Linux samtu 5.3.0-29-generic #31-Ubuntu SMP Fri Jan 17 17:27:26 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux
nodejs-express/ex (pr/604 *$%) % docker -v
Docker version 19.03.6, build 369ce74a3c
```

Since user-app is being mounted in from the host, I think running with the host user's credentials makes a lot of sense... if it can be made to work.

**Comment by skoh7645** [comment]
@sam-github Your image may be using the previously built cached layers. You might want to try removing all docker images before running `appsody stack package` and getting a clean build.

Additionally, I don't think adding `ENV APPSODY_USER_RUN_AS_LOCAL=true` to `image/Dockerfile-stack` will be enough to make `nodejs-express` run as a non-root user.

I think the most comprehensive solution for making the node stacks run as a non-root user is the changes in this PR: https://github.com/appsody/stacks/pull/624. I am using the host's ID as the container's ID and pointing all files that need to be written to, to `/project/user-app`, where all files are owned by the host's/container's ID.

File: incubator/nodejs/image/project/Dockerfile [hunk 2]
```diff
@@ -23,15 +25,18 @@ RUN apt-get update \
...
+ && chmod -R g=u /project
 
 WORKDIR "/project/user-app"
 
 ENV NODE_ENV production
 
-USER node
+USER node_user
 
 CMD ["npm", "start"]
```
Comment (sam-github) [inline incubator/nodejs/image/project/Dockerfile]: ^--- missing NL at end of line (GH and git don't like it)

**Comment by sam-github** [comment]
So, the changes in here are now a copy of https://github.com/appsody/stacks/pull/624?

I'm not sure if you want me to review, or still have open questions.

I cleaned all containers and images, packaged the stack, and tried to use it. No luck.

It fails to run, npm has no write access to /project/user-app/node_modules in the container, and appsody caused docker to create a node_modules on the host  that is owned by `root:root` (I'm on Linux, if that's relevant).

Same for you?

```
incubator/nodejs (pr/604 $%) % pwd
/home/sam/w/cloud-node/appsody-stacks/incubator/nodejs
incubator/nodejs (pr/604 $%) % mkdir ex
incubator/nodejs (pr/604 $%) % cd ex
nodejs/ex (pr/604 $%) % appsody init dev.local/nodejs
Checking stack requirements...
Docker requirements met
Appsody requirements met
Running appsody init...
Downloading nodejs template project from file:///home/sam/.appsody/stacks/dev.local/nodejs.v0.3.3.templates.simple.tar.gz
Download complete. Extracting files from /home/sam/w/cloud-node/appsody-stacks/incubator/nodejs/ex/nodejs.tar.gz
Setting up the development environment
Your Appsody project name has been set to ex
Using local cache for image dev.local/appsody/nodejs:0.3
Running command: docker run --rm --entrypoint /bin/bash dev.local/appsody/nodejs:0.3 -c find /project -type f -name .appsody-init.sh
Successfully initialized Appsody project with the dev.local/nodejs stack and the default template.
nodejs/ex (pr/604 $%) % ls
./  ../  app.js  .appsody-config.yaml  .gitignore  package.json  package-lock.json  .vscode/
nodejs/ex (pr/604 $%) % appsody run
Running development environment...
Using local cache for image dev.local/appsody/nodejs:0.3
Running command: docker run --rm -p 3000:3000 -p 9229:9229 --name ex-dev -v /home/sam/w/cloud-node/appsody-stacks/incubator/nodejs/ex/:/project/user-app -v ex-deps:/project/user-app/node_modules -v appsody-controller-0.3.3:/.appsody -t --entrypoint /.appsody/appsody-controller dev.local/appsody/nodejs:0.3 --mode=run
[Container] Running APPSODY_PREP command: npm install && npm audit fix
npm WARN checkPermissions Missing write access to /project/user-app/node_modules/ansi-colors
[Container] npm WARN checkPermissions Missing write access to /project/user-app/node_modules/ansi-regex
[Container] npm WARN checkPermissions Missing write access to /project/user-app/node_modules/assert-plus
.......... etc.....
[Container] npm ERR!   errno: -13,
[Container] [Error] FATAL error APPSODY_PREP command received an error.  The controller is exiting: exit status 243
[Error] Error in 'appsody run': exit status 1
nodejs/ex (pr/604 $%) % ls -l 
total 36
drwxrwxr-x 4 sam  sam  4096 Feb 24 11:54 ./
drwxrwxr-x 5 sam  sam  4096 Feb 24 11:54 ../
-rw-rw-r-- 1 sam  sam    37 Feb 24 11:54 app.js
-rw-rw-r-- 1 sam  sam    53 Feb 24 11:54 .appsody-config.yaml
-rw-rw-r-- 1 sam  sam  1316 Feb 24 11:54 .gitignore
drwxr-xr-x 2 root root 4096 Feb 24 11:54 node_modules/
-rw-rw-r-- 1 sam  sam   358 Feb 24 11:54 package.json
-rw-rw-r-- 1 sam  sam    76 Feb 24 11:54 package-lock.json
drwxr-xr-x 2 sam  sam  4096 Feb 24 11:54 .vscode/
```

**Comment by skoh7645** [comment]
@sam-github you are getting `node_modules` owned by `root:root` because it is reusing the same docker volume for caching dependencies. If you delete the docker volume (in this case, `ex-deps`) and run `appsody run`, the volume should be recreated with the right file permissions. 
I mentioned the `node-red` PR because adding some of those changes to this PR seems to fix file permission issues with writing to the `resourceRegistry.log` file
```
APPSODY_USER_RUN_AS_LOCAL=true
ENV NODE_PATH=/project/user-app/node_modules
ENV NPM_CONFIG_CACHE=/project/user-app/.npm
ENV XDG_CONFIG_HOME=/project/user-app/.config
```
As stack maintainer, would you prefer this PR goes in as is (with the warning that you cannot access this log file - might be cleaner than using the host's ID in the container and pointing files to different directories), or if you test adding those changes and it works, would you prefer to include it in the PR?

**Comment by neeraj-laad** [comment]
As a general rule, I prefer to avoid using APPSODY_RUN_AS_LOCAL as much as possible as that is just an indication that we are writing to user's filesystem from within the container and that is not a great idea.

**Comment by sam-github** [comment]
<details><summary>This PR does not run for me. See below. I ran `docker system prune --all --volumes --force` before to get a clean system state, and don't think I'm seeing a volume persistence issue because I had no docker volumes before doing `appsody run`.</summary>

```
incubator/nodejs (pr/604 $%) % appsody stack package
******************************************
Running appsody stack package
******************************************
[Warning] Not all labels will be set. Unable to determine origin to compute repository URL
Running docker build
Running command: docker build -t dev.local/appsody/nodejs:0.3.3 -t dev.local/appsody/nodejs:0.3 -t dev.local/appsody/nodejs:0 -t dev.local/appsody/nodejs --label org.opencontainers.image.created=2020-02-25T09:58:56-08:00 --label org.opencontainers.image.authors=Sam Roberts <vieuxtech@gmail.com> --label org.opencontainers.image.version=0.3.3 --label org.opencontainers.image.licenses=Apache-2.0 --label org.opencontainers.image.title=Node.js --label org.opencontainers.image.description=Runtime for Node.js applications --label dev.appsody.stack.id=nodejs --label dev.appsody.stack.tag=dev.local/appsody/nodejs:0.3.3 -f /home/sam/.appsody/stacks/packaging-nodejs/image/Dockerfile-stack /home/sam/.appsody/stacks/packaging-nodejs/image
[Docker] Sending build context to Docker daemon  22.53kB
[Docker] Step 1/37 : FROM node:12
target is: /home/sam/.appsody/stacks/dev.local/nodejs.v0.3.3.templates.
...
new target is: /home/sam/.appsody/stacks/dev.local/nodejs.v0.3.3.templates.simple.tar.gz
Writing: /home/sam/.appsody/stacks/dev.local/dev.local-index.yaml
Your local stack is available as part of repo dev.local
incubator/nodejs (pr/604 $%) % rm -rf ../ex
incubator/nodejs (pr/604 $) % mkdir ../ex
incubator/nodejs (pr/604 $) % cd ../ex
incubator/ex (pr/604 $) % appsody init dev.local/nodejs
Checking stack requirements...
Docker requirements met
Appsody requirements met
Running appsody init...
Downloading nodejs template project from file:///home/sam/.appsody/stacks/dev.local/nodejs.v0.3.3.templates.simple.tar.gz
Download complete. Extracting files from /home/sam/w/cloud-node/appsody-stacks/incubator/ex/nodejs.tar.gz
Setting up the development environment
Your Appsody project name has been set to ex
Using local cache for image dev.local/appsody/nodejs:0.3
Running command: docker run --rm --entrypoint /bin/bash dev.local/appsody/nodejs:0.3 -c find /project -type f -name .appsody-init.sh
Successfully initialized Appsody project with the dev.local/nodejs stack and the default template.
incubator/ex (pr/604 $%) % docker volume ls
DRIVER              VOLUME NAME
incubator/ex (pr/604 $%) % appsody run
Running development environment...
Using local cache for image dev.local/appsody/nodejs:0.3
Running command: docker run --rm -v appsody-controller-0.3.3:/.appsody appsody/init-controller:0.3.3
Unable to find image 'appsody/init-controller:0.3.3' locally
0.3.3: Pulling from appsody/init-controller
322973677ef5: Pulling fs layer
d6bf415d0a7d: Pulling fs layer
322973677ef5: Verifying Checksum
322973677ef5: Download complete
d6bf415d0a7d: Download complete
322973677ef5: Pull complete
d6bf415d0a7d: Pull complete
Digest: sha256:038374ea014798d9cf1ad8f591eb0985039b7edaeb638f2cdc540747262ee7ed
Status: Downloaded newer image for appsody/init-controller:0.3.3
Running command: docker run --rm -p 3000:3000 -p 9229:9229 --name ex-dev -v /home/sam/w/cloud-node/appsody-stacks/incubator/ex/:/project/user-app -v ex-deps:/project/user-app/node_modules -v appsody-controller-0.3.3:/.appsody -t --entrypoint /.appsody/appsody-controller dev.local/appsody/nodejs:0.3 --mode=run
[Container] Running APPSODY_PREP command: npm install && npm audit fix
up to date in 0.332s
[Container] found 0 vulnerabilities
[Container]
[Container] up to date in 0.037s
[Container] fixed 0 of 0 vulnerabilities in 0 scanned packages
[Container] Running command:  npm start --node-options --require=appmetrics-dash/attach
[Container]
[Container] > nodejs-simple@0.1.0 start /project/user-app
[Container] > node app.js
[Container]
[Container] [Tue Feb 25 18:03:20 2020] com.ibm.diagnostics.healthcenter.loader INFO: Node Application Metrics 5.1.1.202002251800 (Agent Core 4.0.5)
[Container] Hello from Node.js 12!
[Container] log4js.fileAppender - Writing to file resourceRegistry.log, error happened  [Error: EACCES: permission denied, open '/project/user-app/resourceRegistry.log'] {
[Container]   errno: -13,
[Container]   code: 'EACCES',
[Container]   syscall: 'open',
[Container]   path: '/project/user-app/resourceRegistry.log'
[Container] }
```
</details>

**Comment by sam-github** [comment]
Was it your intention to not use the `node` user, and instead create a new one? The mixture of permissions seems unusual, and may be the source of the perm problems, because /project/user-app is owned by `node`, but `npm start` is being run as the user `node_user`:
```
incubator/nodejs (pr/604 $%) % docker exec -it f042030663b7 /bin/bash
node_user@f042030663b7:/project/user-app$ ls
app.js  node_modules  package-lock.json  package.json
node_user@f042030663b7:/project/user-app$ ls -l
total 16
-rw-rw-r-- 1 node      node        37 Feb 25 18:03 app.js
drwxr-xr-x 2 node_user node_user 4096 Feb 25 17:59 node_modules
-rw-rw-r-- 1 node      node        76 Feb 25 18:03 package-lock.json
-rw-rw-r-- 1 node      node       358 Feb 25 18:03 package.json
node_user@f042030663b7:/project/user-app$ id    
uid=1001(node_user) gid=1001(node_user) groups=1001(node_user)
node_user@f042030663b7:/project/user-app$ touch fu
touch: cannot touch 'fu': Permission denied
node_user@f042030663b7:/project/user-app$ ls -ld /project /project/user-app
drwxr-xr-x 1 node_user node_group 4096 Feb 25 18:00 /project
drwxrwxr-x 4 node      node       4096 Feb 25 18:03 /project/user-app
```

**Comment by sam-github** [comment]
@skoh7645 

> As stack maintainer, would you prefer this PR goes in as is

I don't have strong opinions, as long as it runs (which it doesn't now, not for me).

>  (with the warning that you cannot access this log file - might be cleaner than using the host's ID in the container and pointing files to different directories),

Not sure what you mean by "pointing files to different directories"


>  or if you test adding those changes and it works, would you prefer to include it in the PR?

"those changes" ... those refers to #624?

"include it in the PR" ... it is what?


@neeraj-laad 

> As a general rule, I prefer to avoid using APPSODY_RUN_AS_LOCAL as much as possible as that is just an indication that we are writing to user's filesystem from within the container and that is not a great idea.

For production, that would be a terrible idea, but appsody is trying to build a local dev experience with files being edited live on the host, and the app and unit tests running in a container, with the host FS mapped into the container...

Its not unreasonable for an app to write to FS, particularly in a "dev mode". Templates might get recompiled on the fly, unit tests will certainly write to FS in all kinds of configurations, such as when collecting coverage information, or creating text fixtures.

This doesn't strike me as a Node.js specific issue, that would be true of most/all runtimes, wouldn't it?

For example, with this PR, with a pretend test script that fakes writing coverage, this won't run:
```
incubator/ex (pr/604 $%) % vi package.json 
incubator/ex (pr/604 $%) % appsody test   
Running test environment
Using local cache for image dev.local/appsody/nodejs:0.3
Running command: docker run --rm -p 3000:3000 -p 9229:9229 --name ex-dev -v /home/sam/w/cloud-node/appsody-stacks/incubator/ex/:/project/user-app -v ex-deps:/project/user-app/node_modules -v 
appsody-controller-0.3.3:/.appsody -t --entrypoint /.appsody/appsody-controller dev.local/appsody/nodejs:0.3 --mode=test
[Container] Running APPSODY_PREP command: npm install && npm audit fix
up to date in 0.508s
[Container] found 0 vulnerabilities
[Container] 
up to date in 0.04s
[Container] fixed 0 of 0 vulnerabilities in 0 scanned packages
[Container] Running command:  npm test
[Container] 
[Container] > nodejs-simple@0.1.0 test /project/user-app
[Container] > touch coverage.txt
[Container] 
[Container] touch: cannot touch 'coverage.txt': Permission denied
[Container] npm ERR! Test failed.  See above for more details.
[Container] [Error] Wait received error with status code: 1 due to error: exit status 1
[Error] Error in 'appsody test': exit status 1
incubator/ex (pr/604 $%) % git diff
incubator/ex (pr/604 $%) % grep test package.json 
    "test": "touch coverage.txt",
```

Anyhow, I originally got pulled into this to answer some questions about IBM's apm plugin and its unfortunate writing of resource.log, which can't be disabled. I filed an issue upstream (no response yet on that).

In terms of generally how appsody will enable the code running in the container to write to its src directory, I'm not sure.

I suppose it might be possible to make (and document) "rules" -- such as "apps can't run in appsody unless they always write not to the CWD, but to `APPSODY_TMP`, or maybe appsody should actually run with the CWD in /tmp? I'm not sure.

**Comment by Kamran64** [comment]
@sam-github - I initially used the `node` user that comes with the base image but was running into permissions problems (because of root owned directories). I found that when I set up a new user (`node_user`) and ran the stack, directories were now owned by the new user instead of root. I'm assumed there are some constraints set on this user. I wasn't expecting files to be owned by `node` user at all in this PR now unless the default user in the image is `node` instead of `root` but I hadn't observed that either when testing before. However, I'll take another look since this PR has been opened for a while so I might have missed something.

**Comment by sam-github** [comment]
Needs to be synced with the other PRs: #589 #604 #624

**Comment by sam-github** [comment]
Talked to @skoh7645 , it sounds like #624 is the approach that will be taken with all 3.

**Comment by Kamran64** [comment]
Let's close this PR for now. We need to come up with a basis for how we should approach this problem. It looks like the changes to https://github.com/appsody/stacks/pull/624 might be the way forward. We can reopen this later once the implementation details are finalised.

**Pull request closed without merging**